src/v/utils/vint.h

```diff
@@ -12,8 +12,111 @@
 #pragma once
 #include "bytes/bytes.h"
 
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
 #include <cstdint>
 
+namespace unsigned_vint {
+/// At most 5 bytes are needed to encode a 32 bit value
+inline constexpr size_t max_length = 5;
+/// Consume up to 5 iterations (0-4) of reading 7 bits each time
+constexpr auto limit = ((max_length - 1) * 7);
+
+namespace detail {
+struct var_decoder {
+    uint64_t result = 0;
+    size_t bytes_read = 0;
+    uint64_t shift = 0;
+    uint8_t limit;
+
+    explicit var_decoder(uint8_t limit)
+      : limit(limit) {}
+
+    /**
+     * @brief Incorporate a new byte into the state.
+     * Returns true if the outer loop should break (last byte found).
+     */
+    bool accept(uint8_t byte) {
+        /// \brief almost identical impl to leveldb, made generic for c++
+        /// friendliness
+        /// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
+        bytes_read++;
+        if (byte & 128) {
+            result |= ((byte & 127) << shift);
+        } else {
+            result |= byte << shift;
+            return true;
+        }
+        shift += 7;
+        return shift > limit;
+    }
+};
+
+template<typename Range>
+inline std::pair<uint64_t, size_t>
+deserialize(Range&& r, uint8_t limit) noexcept {
+    var_decoder decoder(limit);
+    for (auto byte : r) {
+        if (decoder.accept(static_cast<uint8_t>(byte))) {
+            break;
+        }
+    }
+    return std::make_pair(decoder.result, decoder.bytes_read);
+}
+
+} // namespace detail
+
+inline ss::future<std::pair<uint32_t, size_t>>
```

Comment:
Keeping this method in the header results in strange compiler output like this:
```
[448/634] Building CXX object src/v/kafka/CMakeFiles/v_kafka.dir/server/protocol_utils.cc.o
<unknown>:0: error: Cannot represent a difference across sections
<unknown>:0: error: Cannot represent a difference across sections
<unknown>:0: error: Cannot represent a difference across sections
```
Will move to a source file